src/codewhisperer/language/codewhispererProgrammingLanguage.ts

```diff
@@ -0,0 +1,161 @@
+import { CodeWhispererCpp } from './codewhispererCpp'
+import { CodeWhispererGo } from './codewhispererGo'
+import { CodeWhispererKotlin } from './codewhispererKotlin'
+import { CodeWhispererPhp } from './codewhispererPhp'
+import { CodeWhispererRuby } from './codewhispererRuby'
+import { CodeWhispererRust } from './codewhispererRust'
+import { CodeWhispererScala } from './codewhispererScala'
+import { CodeWhispererShellscript } from './codewhispererShellscript'
+import { CodeWhispererSql } from './codewhispererSql'
+import { CodewhispererLanguage } from '../../shared/telemetry/telemetry.gen'
+import path from 'path'
+
+export function getLanguage(document: vscode.TextDocument): CodeWhispererProgrammingLanguage {
+    const languageId = document.languageId
+    return (
+        platformLanguageMap[languageId as PlatformLanguageId] ??
+        extToLanguage.get(path.extname(document.fileName)) ??
+        platformLanguageMap.plaintext
+    )
+}
+
+/**
+ * This is for notebook files map to a new filename with the corresponding language extension
+ * @param platformLanguageId : official vscode languageId
+ * @returns corresponding language extension if any, otherwise undefined
+ */
+export function getLanguageExtensionForNotebook(platformLanguageId: string): string | undefined {
+    const lang = platformLanguageMap[platformLanguageId as PlatformLanguageId] ?? undefined
+    if (!lang) {
+        return undefined
+    }
+
+    const ext: string | undefined = languageToExts[lang.id][0]
+    return ext
+}
+
+export abstract class CodeWhispererProgrammingLanguage {
+    abstract readonly id: CodewhispererLanguage
+
+    abstract toCodeWhispererRuntimeLanguage(): CodeWhispererProgrammingLanguage
+
+    abstract isCodeCompletionSupported(): boolean
+
+    abstract isCodeScanSupported(): boolean
+
+    static from(id: string): CodeWhispererProgrammingLanguage {
+        return platformLanguageMap[id as PlatformLanguageId] ?? platformLanguageMap.plaintext
+    }
+}
+
+class CodeWhispererPlaintext extends CodeWhispererProgrammingLanguage {
+    readonly id = 'plaintext'
+    toCodeWhispererRuntimeLanguage(): CodeWhispererProgrammingLanguage {
+        return this
+    }
+
+    isCodeCompletionSupported(): boolean {
+        return false
+    }
+
+    isCodeScanSupported(): boolean {
+        return false
+    }
+}
+
+type PlatformLanguageId = (typeof platformLanguageIds)[number]
+
+const platformLanguageIds = [
+    'java',
+    'python',
+    'javascript',
+    'javascriptreact',
+    'typescript',
+    'typescriptreact',
+    'csharp',
+    'c',
+    'cpp',
+    'c_cpp', // Cloud9 reports C++ files with this language-id.
+    'go',
+    'kotlin',
+    'php',
+    'ruby',
+    'rust',
+    'scala',
+    'shellscript',
+    'sh', // Cloud9 reports bash files with this language-id
+    'sql',
+    'plaintext',
+] as const
+
+const platformLanguageMap: Record<PlatformLanguageId, CodeWhispererProgrammingLanguage> = {
+    java: new CodeWhispererJava(),
```

Comment:
@justinmk3, a naive q, any suggestion on how to resolve "circular dependency" here?
```
 src/codewhisperer/language/languages/codewhispererPython.ts -> src/codewhisperer/language/codewhispererProgrammingLanguage.ts -> src/codewhisperer/language/languages/codewhispererPython.ts
```

Reply:
I only know it can be resolved if I put all "languegages" in the same file as its base class "CodeWhispererProgrammingLanguage"